Issue 383 - Implement User Group main page

mreynolds389 opened the pull request:
Create the main page and the add/delete modals for User Groups

relates: https://github.com/freeipa/freeipa-webui/issues/383

carma12 commented on src/navigation/AppRoutes.tsx:
```diff
@@ -60,6 +61,7 @@ export const AppRoutes = (): React.ReactElement => (
       </Route>
       <Route path="user-groups">
         <Route path="" element={<UserGroups />} />
+        <Route path="settings" element={<UserGroupsTabs />} />
```
Just FYI (no need to change anything): This format will change in the future to use `:userGroupId` instead of `/settings`.  This will allow to reference the Settings page of a give entity page (users, hosts, services, user groups, etc.) in a more logical and consistent way using a specific ID, instead of relying on a generalist name as it is `settings`. 

E.g., instead of `<BASE_URL>/user-groups/settings` it will be `<BASE_URL>/user-groups/<userGroupId>`.

But it is ok to keep the format you added here from now :)

carma12 commented on src/pages/UserGroups/UserGroups.tsx:
```diff
@@ -1,8 +1,556 @@
+            groupsList.push(groupsListResult[i].result);
+          }
+
+          // Update 'Hosts' slice data
```
Reword: `Hosts` --> `User groups`.

carma12 commented on src/pages/UserGroups/UserGroups.tsx:
```diff
@@ -1,8 +1,556 @@
+      element: (
+        <SearchInputLayout
+          name="search"
+          ariaLabel="Search hosts"
```
Reword: `hosts` --> `user groups`.

carma12 commented on src/pages/UserGroups/UserGroupsTabs.tsx:
```diff
@@ -0,0 +1,107 @@
+// Other
+// import HostsSettings from "./HostsSettings";
+// import HostsMemberOf from "./HostsMemberOf";
+// import HostsManagedBy from "./HostsManagedBy";
```
Leftovers?

carma12 commented on src/pages/UserGroups/UserGroupsTabs.tsx:
```diff
@@ -0,0 +1,107 @@
+  const pagesVisited = [
+    {
+      name: "User Groups",
+      url: URL_PREFIX + "/usergroups",
```
Typo: `usergroups`--> `user-groups`

carma12 commented on src/store/Identity/userGroups-slice.ts:
```diff
@@ -1,23 +1,41 @@
 }
 
 const initialState: UserGroupState = {
-  userGroupList: userGroupsJson,
```
Since the `userGroupsJson` file containing the dummy data won't be used anymore, I think it is safe to remove it.

carma12 reviewed (changes requested):
First round of review, please check the comments below.

Apart from that, there are some suggestions / questions related to this PR:
- Does this solution includes the search input functionality? I have tried it and doesn't seems to work for me.
- I think it would be nice to return to page 1 when after an add/remove operation.
- When adding multiple elements, those seem unsorted. So an enhancement could be to sort the list after the group elements are retrieved, added, or removed.
- I think the tests for this section can be reviewed by @miskopo as he could provide a more insightful point of view as QE ;)

mreynolds389 commented on src/components/modals/AddUserGroup.tsx:
```diff
@@ -0,0 +1,394 @@
+    setGroupType(value);
+  };
+
+  console.log("MAKR group type: ", groupType);
```
> Is this intended or a leftover?

I thought I removed that one :-)

mreynolds389 commented:
@carma12 I applied all the changes, but I could not reproduce the table getting out of order.  Adding new user groups are added and placed in the correct table order

carma12 commented:
> @carma12 I applied all the changes, but I could not reproduce the table getting out of order. Adding new user groups are added and placed in the correct table position

That's odd - I tried it yesterday and when I inserted new entries the list was unsorted. Today I have tried it again and works fine. Maybe it was a cache problem from my side :) 

But I noticed something else that I want to double check: when inserting a new entry that alphabetically should be in other pages that is not page 1, the new entry is not there when navigate to the x page. Steps to reproduce it:
1.  Create a user group called `xyz-usergroup` (should be in the last pages when created)
2. Navigate to the last page or where the item shoud be located
3. The newly-created item is not there. Only when hit the 'Refresh' button the element is shown (this doesn't happen all times)

Can you reproduce the same behavior? Otherwise, I can blame my messy cache :)

carma12 commented:
> > > @carma12 I applied all the changes, but I could not reproduce the table getting out of order. Adding new user groups are added and placed in the correct table position
> > 
> > 
> > That's odd - I tried it yesterday and when I inserted new entries the list was unsorted. Today I have tried it again and works fine. Maybe it was a cache problem from my side :)
> > But I noticed something else that I want to double check: when inserting a new entry that alphabetically should be in other pages that is not page 1, the new entry is not there when navigate to the x page. Steps to reproduce it:
> > ```
> > 1. Create a user group called `xyz-usergroup` (should be in the last pages when created)
> > 
> > 2. Navigate to the last page or where the item shoud be located
> > 
> > 3. The newly-created item is not there. Only when hit the 'Refresh' button the element is shown (this doesn't happen all times)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Can you reproduce the same behavior? Otherwise, I can blame my messy cache :)
> 
> I'm reproducing a similar behavior that seems to affect all the pages' pagination :-(
> 
> I start adding groups (or hosts, etc) and once I get to 11 entries then I added a group named zzzzzz. I goto to page 2, I can it loading the data, and the content is correct. Then I go back to page 1 and add zzzzz2 and zzzzz3. On page 1 it correctly shows that there are 14 entries. When I go back to page 2 (it does not reload) it is not updated. Pagination says there are only 12 entries and it's the old content.
> 
> Since issue seems to affect all pages I think we should open a new ticket for it. WDYT?

That is a good idea, feel free to open it. 

I think this might be related to the fact that the information of the entity (user group in this case) is not re-fetched once an add is done. And considering that we are fetching only the data that will be shown in a given page, this issue is more prone to happen.

mreynolds389 commented:
@carma12  Fixed the last issue with page sets being outdated.  I was going to open a new ticket but it turns out it only impacted hosts and services.  Since the fix is quite simple I just included it in this PR.

carma12 reviewed (changes requested):
It seems that the problem is still there and affects adding and removing buttons. Please find the attached video:  https://drive.google.com/file/d/1G1ZfHcsKhKNDuRsrdD7LdKUT2BEoGS04/view?usp=sharing

mreynolds389 commented:
> It seems that the problem is still there and affects adding and removing buttons. Please find the attached video: https://drive.google.com/file/d/1G1ZfHcsKhKNDuRsrdD7LdKUT2BEoGS04/view?usp=sharing

Ah ok, fixed now.  I had fixed and then tweaked it and broke it :-(  Errrr.  Anyway it's working now
